src/repl/scala/tools/nsc/interpreter/Interface.scala

```diff
@@ -0,0 +1,283 @@
+  def valueOfTerm(id: String): Option[Any]
+}
+
+/**
+  * The interface used to expose the repl as a Java Script Engine
+  */
+trait ScriptedRepl extends Repl {
+  def compile(code: String): Boolean
+  def compile(line: String, synthetic: Boolean): Either[Result, Request]
+
+  def call(name: String, args: Any*): Either[Throwable, AnyRef]
+
+  // TODO: should we move more of the wrapping from shell/Scripted to interpreter/Scripted,
+  // to avoid exposing this low-level stuff to the frontend (shell)?
+  def evalName: String
+  def evalPath: String
+
+  def recordRequest(req: Request): Unit
+  def addBackReferences(req: Request): Either[String, Request]
+}
+
+trait ReplReporter extends Reporter {
+  def out: PrintWriter
+
+  /**
+    * Print message (info/warning/error).
+    * By default, messages beyond a certain length are truncated (see `withoutTruncating`),
+    * and internal repl wrapping is removed (see `withoutUnwrapping` and `unmangleInterpreterOutput`).
+    * To suppress all output, use `suppressOutput`
+    */
+  def printMessage(msg: String): Unit
+
+  /** Don't print any errors/messages/echos during the execution of `body`.
+    */
+  def suppressOutput[T](body: => T): T
+
+  /** Suppress truncation during the executing of `body`.
+    */
+  def withoutTruncating[T](body: => T): T
+
+  /** Do not remove interpreter wrappers ($iw etc) from all output during the execution of `body`.
+    */
+  def withoutUnwrapping(body: => Unit): Unit
+
+
+  /** Print result (Right --> success, Left --> error)
+    */
+  def printResult(result: Either[String, String]): Unit
+
+  /** Don't print result lines.
+    */
+  def withoutPrintingResults[T](body: => T): T
+
+  /** Whether we're printing results (should only be used from the shell).
+    */
+  def printResults: Boolean
+
+  /** Toggle whether to print results (should only be used from the shell).
+    */
+  def togglePrintResults(): Unit
+
+
+  //// println debugging ftw
+  def isDebug: Boolean
+  def debug(msg: => String): Unit = if (isDebug) echo(msg)
+
+  def isTrace: Boolean
+  def trace(msg: => String): Unit = if (isTrace) echo(msg)
+
+  //// Internal signalling from repl to shell
+
+  /** Currently executing request (used to determine position of error in terms of user-submitted code)
+    *
+    * TODO: should no longer be needed if we do wrapping after type checking
+    */
+  def currentRequest: ReplRequest
+
+  /** Set currently executing request.
+    */
+  def currentRequest_= (req: ReplRequest): Unit
+}
+
+trait ReplRequest {
+  def line: String
+  def originalLine: String
+  def preambleEndPos: Position
+
+  def eval: Either[Throwable, AnyRef]
+}
+
+/**
+  *
+  * Created by scala.tools.nsc.interpreter.Repl#presentationCompile
+  */
+trait PresentationCompilationResult {
+  protected def cursor: Int
+  protected def buf: String
+
+  def cleanup(): Unit
+
+  def print: String
+
+  def typeAt(start: Int, end: Int): String
+
+  def candidates(tabCount: Int): (Int, List[String])
```

Comment:
This is somewhat REPL specific, and it seems odd to already be String-ifiying the types at this level of API. I'm happy to proceed with the refactoring and refine later, though.